rfcs/001-jwt-service-to-service-authentication.md

```diff
@@ -0,0 +1,280 @@
+the likelihood of scaling and management issues for clients.
+- Integration barriers: A lack of standardisation drives up integration effort, blocks reuse of shared client libraries,
+and increases time-to-value for consuming APIs.
+
+## Background
+
+- JSON Web Token (JWT): A compact, URL-safe token format used for transmitting claims between parties, often signed
+with asymmetric keys, see [RFC 7519](https://datatracker.ietf.org/doc/html/rfc7519).
+- Decentralised Identifier (DID): A W3C standard for self-sovereign identifiers, resolved via `/.well-known/did.json`
+or other DID methods, see [W3C DID Core Specification](https://www.w3.org/TR/did-core/).
+- JSON Web Key (JWK): A JSON representation of cryptographic keys, see [RFC 7517](https://datatracker.ietf.org/doc/html/rfc7517)
+- DPoP (Demonstrating Proof of Possession): An OAuth 2.0 extension for binding tokens to sender keys, preventing replay,
+see [RFC 7800](https://datatracker.ietf.org/doc/html/rfc7800).
+
+## Proposal
+
+This RFC defines a service-to-service authentication mechanism using:  
+
+1. **DIDs as Service Identities**
+
+    - Each service SHOULD have a DID (e.g. `did:web:forms.service.gov.uk`) and expose a DID document with one or more
+    verification methods.  
+
+1. **JWTs for Proof-of-Possession**
+
+    - API requests MUST include a JWT, either as:
+        - Bearer token - [RFC 6750](https://www.rfc-editor.org/rfc/rfc6750#section-2.1)
+        - `DPoP` header - [RFC 9449](https://datatracker.ietf.org/doc/html/rfc9449)
+    - The JWT header MUST include:
+        - `typ`: A field with the value `dpop+jwt` or `JWT`.
+        - `alg`: An identifier for a JWS asymmetric digital signature algorithm.
+    - The JWT header MAY include:
+        - `jwk`: Represents the public key chosen by the client in JSON Web Key.
+        - `kid`: A hint indicating which key the client used to generate the token signature.
+    - The JWT payload MUST include at least:
+        - `jti`: Unique identifier for the JWT.
+        - `iat`: Issued-at timestamp (UNIX time).
+    - The JWT payload MUST include either:
+        - `htu` and `htm`
+            - `htu`: The HTTP target URI of the API call (e.g. `https://api.notifications.service.gov.uk/v2/notifications/email`).
+            - `htm`: The HTTP method (e.g. `POST`).  
+
+        or
+        - `aud`: The audience as scheme, authority and path (heirarchical part) (e.g. `https://api.notifications.service.gov.uk/v2/notifications/email`)
+                or DID urn (e.g. `did:web:api.notifications.service.gov.uk:v2:notifications`).
+    - The JWT payload SHOULD include:
+        - `iss`: The client as scheme, authority and path (heirarchical part) (e.g. `https://forms.service.gov.uk/api/`)
+                or DID urn (e.g. `did:web:forms.service.gov.uk%3A443:api`).
+        - `exp`: Expiry time (UNIX time).
+        - `cnf`: A JSON object and the members of that object identify the proof-of-possession key.
```

Comment:
It would be good to have `cnf` in one the examples; it’s not clear to me what should be in there and how it should be used.

Reply:
It's [part of RFC 7800](https://datatracker.ietf.org/doc/html/rfc7800#section-3.1) but I must admit I'm not entirely convinced on the use given for DPoP `jwk` is expected to be set in the header.